src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -510,14 +562,180 @@ internal void ReadToken(string encodedJson)
+        /// <exception cref="SecurityTokenMalformedException">if <paramref name="encodedTokenMemory"/> has more than 4 dots.</exception>
+        internal void ReadToken(ReadOnlyMemory<char> encodedTokenMemory)
+        {
+            // JWT must have 2 dots for JWS or 4 dots for JWE (a.b.c.d.e)
+            ReadOnlySpan<char> encodedTokenSpan = encodedTokenMemory.Span;
+
+            Dot1 = encodedTokenSpan.IndexOf("."); 
+            if (Dot1 == -1 || Dot1 == encodedTokenSpan.Length - 1)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14100));
+
+            Dot2 = encodedTokenSpan.Slice(Dot1 + 1).IndexOf(".");
+            if (Dot2 == -1)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14120));
+
+            Dot2 = Dot1 + Dot2 + 1;
+            Dot3 = (Dot2 == encodedTokenSpan.Length - 1) ? -1 : encodedTokenSpan.Slice(Dot2 + 1).IndexOf('.');
+
+            if (Dot3 == -1)
+            {
+                // JWS has two dots
+                // JWS: https://www.rfc-editor.org/rfc/rfc7515
+                // Format: https://www.rfc-editor.org/rfc/rfc7515#page-7
+
+                IsSigned = !(Dot2 + 1 == encodedTokenSpan.Length);
+                try
+                {
+                    Header = CreateClaimSet(encodedTokenSpan, 0, Dot1, createHeaderClaimSet: true);
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(
+                        LogMessages.IDX14102,
+                        LogHelper.MarkAsUnsafeSecurityArtifact(encodedTokenSpan.Slice(0, Dot1).ToString(), t => t.ToString())), // TODO: Add an overload to LogHelper.MarkAsUnsafeSecurityArtifact that accepts span?
+                        ex));
+                }
+
+                try
+                {
+                    Payload = CreateClaimSet(encodedTokenSpan, Dot1 + 1, Dot2 - Dot1 - 1, createHeaderClaimSet: false);
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(
+                        LogMessages.IDX14101,
+                        LogHelper.MarkAsUnsafeSecurityArtifact(encodedTokenSpan.Slice(Dot1 + 1, Dot2).ToString(), t => t.ToString())),
+                        ex));
+                }
+            }
+            else
+            {
+                // JWE: https://www.rfc-editor.org/rfc/rfc7516
+                // Format: https://www.rfc-editor.org/rfc/rfc7516#page-8
+                // empty payload for JWE's {encrypted tokens}.
+                Payload = new JsonClaimSet();
+
+                if (Dot3 == encodedTokenSpan.Length) // TODO: Should this be encodedJsonSpan.Length - 1? Using encodedJsonSpan.Length doesn't add up
```

Comment:
>Usin [](http://example.com/codeflow?start=105&length=4)

Is it because if `Dot3` is equal to `encodedTokenSpan.Length - 1`, it means that the third dot is the second-to-last character in the encoded token, which is an error?

Reply:
My understanding is encodedTokenSpan.Length - 1 represents the last character and index of encodedTokenSpan.Length  would fall outside the range of the Span. That's why I'm wondering if this condition should be rewritten to `Dot3 == encodedTokenSpan.Length - 1`

Reply:
looks like we need a bit more test coverage on this line too